src/plugins/elasticsearch_ui_shared/static/forms/hook_form_lib/hooks/use_field.ts

```diff
@@ -0,0 +1,427 @@
+      }
+      let validationResult;
+
+      try {
+        validationResult = validator({
+          value: (valueToValidate as unknown) as string,
+          errors: validationErrors,
+          formData,
+          path,
+        });
+
+        if (validationResult && exitOnFail !== false) {
+          throw validationResult;
+        }
+      } catch (error) {
+        // If an error is thrown, skip the rest of the validations
+        skip = true;
+        validationResult = error;
+      }
+
+      return validationResult;
+    };
+
+    // Execute each validations for the field sequencially
+    validations.forEach(validation => {
+      const validationResult = runValidation(validation);
+
+      if (validationResult) {
+        const error = getValidationErrorWithMessage(validation, validationResult);
+        validationErrors.push({
+          ...error,
+          validationType: validation.type || VALIDATION_TYPES.FIELD,
+        });
+      }
+    });
+
+    return validationErrors;
+  };
+
+  const validateAsync = async ({
+    formData,
+    value: valueToValidate,
+    validationTypeToValidate,
+  }: {
+    formData: any;
+    value: unknown;
+    validationTypeToValidate?: string;
+  }): Promise<ValidationError[]> => {
+    const validationErrors: ValidationError[] = [];
+    let skip = false;
+
+    // By default, for fields that have an asynchronous validation
+    // we will clear the errors as soon as the field value changes.
+    clearErrors([VALIDATION_TYPES.FIELD, VALIDATION_TYPES.ASYNC]);
+
+    const runValidation = async ({
+      validator,
+      exitOnFail,
+      type: validationType = VALIDATION_TYPES.FIELD,
+    }: ValidationConfig) => {
+      if (
+        skip ||
+        (typeof validationTypeToValidate !== 'undefined' &&
+          validationType !== validationTypeToValidate)
+      ) {
+        return;
+      }
+      let validationResult;
+
+      try {
+        validationResult = await validator({
+          value: (valueToValidate as unknown) as string,
+          errors: validationErrors,
+          formData,
+          path,
+        });
+
+        if (validationResult && exitOnFail !== false) {
+          throw validationResult;
+        }
+      } catch (error) {
+        // If an error is thrown, skip the rest of the validations
+        skip = true;
+        validationResult = error;
+      }
+
+      return validationResult;
+    };
+
+    // Sequencially execute all the validations for the field
+    await validations.reduce(
+      (promise, validation) =>
+        promise.then(async () => {
+          const validationResult = await runValidation(validation);
+
+          if (validationResult) {
+            const error = getValidationErrorWithMessage(validation, validationResult);
+            validationErrors.push({
+              ...error,
+              validationType: validation.type || VALIDATION_TYPES.FIELD,
+            });
+          }
+        }),
+      Promise.resolve()
```

Comment:
Very cool pattern! I've never thought of `Promise.resolve()` as the empty case.

I guess this could be slightly less verbose with the imperative alternative:

```ts
for (const validation of validations) {
  const validationResult = await runValidation(validation);
  // ...
````

(but you don't have to implement, just wanted to point out the cool use of `Promise`+`reduce` here.

Reply:
Good idea, I updated with the imperative alternative. 👍